ext/avromatic/encoder.cpp

```diff
@@ -0,0 +1,267 @@
+      std::vector<uint8_t>& destination_buffer = datum.value<std::vector<uint8_t> >();
+      destination_buffer.reserve(num_bytes);
+      destination_buffer.assign(raw_bytes, raw_bytes + num_bytes);
+      break;
+    }
+    case avro::AVRO_INT:
+    {
+      if (datum.logicalType().type() == avro::LogicalType::DATE) {
+        datum.value<int>() = serialize_logical_date(rb_value);
+      } else {
+        datum.value<int>() = NUM2INT(rb_value);
+      }
+      break;
+    }
+    case avro::AVRO_LONG:
+    {
+      if (datum.logicalType().type() == avro::LogicalType::TIMESTAMP_MILLIS) {
+        datum.value<long>() = serialize_logical_timestamp_millis(rb_value);
+      } else if (datum.logicalType().type() == avro::LogicalType::TIMESTAMP_MICROS) {
+        datum.value<long>() = serialize_logical_timestamp_micros(rb_value);
+      } else {
+        datum.value<long>() = NUM2LONG(rb_value);
+      }
+      break;
+    }
+    case avro::AVRO_FLOAT:
+    {
+      datum.value<float>() = (float)RFLOAT_VALUE(rb_value);
+      break;
+    }
+    case avro::AVRO_DOUBLE:
+    {
+      datum.value<double>() = RFLOAT_VALUE(rb_value);
+      break;
+    }
+    case avro::AVRO_BOOL:
+    {
+      datum.value<bool>() = rb_value == Qtrue;
+      break;
+    }
+    case avro::AVRO_NULL:
+    {
+      // Nothing todo here
+      break;
+    }
+    case avro::AVRO_RECORD:
+    {
+      ruby_model_to_datum(rb_value, datum);
+      break;
+    }
+    case avro::AVRO_ENUM:
+    {
+      datum.value<avro::GenericEnum>().set(StringValueCStr(rb_value));
+      break;
+    }
+    case avro::AVRO_ARRAY:
+    {
+      avro::GenericArray& arrayValue = datum.value<avro::GenericArray>();
+      VALUE rb_avromatic_item_type = rb_funcall(rb_avromatic_type, rb_intern("value_type"), 0);
+      const long num_values = rb_array_len(rb_value);
+      arrayValue.value().resize(num_values, avro::GenericDatum(arrayValue.schema()->leafAt(0)));
+      for (long i=0; i<num_values; ++i) {
+        VALUE rb_nested_value = rb_ary_entry(rb_value, i);
+        ruby_value_to_datum(rb_nested_value, rb_avromatic_item_type, arrayValue.value()[i]);
+      }
+      break;
+    }
+    case avro::AVRO_MAP:
+    {
+      avro::GenericMap& mapValue = datum.value<avro::GenericMap>();
+      VALUE rb_avromatic_item_type = rb_funcall(rb_avromatic_type, rb_intern("value_type"), 0);
+      // TODO: Avoid allocating the key array by using rb_hash_foreach
+      VALUE rb_keys = rb_funcall(rb_value, rb_intern("keys"), 0);
+      const long num_keys = rb_array_len(rb_keys);
+      mapValue.value().resize(num_keys, std::pair<std::string, avro::GenericDatum>("", avro::GenericDatum(mapValue.schema()->leafAt(1))));
+      for (long i=0; i<num_keys; ++i) {
+        VALUE rb_key = rb_ary_entry(rb_keys, i);
+        VALUE rb_utf8_key = rb_funcall(rb_key, rb_intern("encode"), 1, rb_id2str(rb_intern("utf-8")));
+        // TODO: Does this work correctly with utf-8 characters?
```

Comment:
I just did some reading on this and, as far as I can tell, the only time a zero byte should appear in UTF-8 is when it's encoding a literal NUL. For all multibyte characters the values start at 127.

I also found that there is `ruby/encoding.h` lib which seems to have ways to set encoding on rb String types directly from C, `rb_encoding_find_index("UTF-8")`, and subsequently `rb_encoding_associate_index(index, rb_key)`. The docs aren't crystal clear though so I'm not sure it does what you want here. I tried using it like so:

```cpp
        VALUE rb_key = rb_ary_entry(rb_keys, i);
        auto utf8 = rb_enc_find_index("UTF-8");
        rb_enc_associate_index(utf8, rb_key);
```

And then the Ruby code complained that I can't modify a frozen Integer, which, I found baffling. I gave up at that point. :-)